src/server/grpc_service.rs

```diff
@@ -161,84 +162,80 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
-                                             mutations,
-                                             req.take_primary_lock(),
-                                             req.get_start_version(),
-                                             options,
-                                             cb);
-            if let Err(e) = res {
-                let status = RpcStatus::new(RpcStatusCode::ResourceExhausted,
-                                            Some(format!("{}", e)));
-                return sink.fail(status).map_err(|e| error!("kv_get failed: {:?}", e)).boxed();
-            }
-            future.map_err(Error::from)
-                .map(|v| {
-                    let mut resp = PrewriteResponse::new();
-                    if let Some(err) = extract_region_error(&v) {
-                        resp.set_region_error(err);
-                    } else {
-                        resp.set_errors(RepeatedField::from_vec(extract_key_errors(v)));
-                    }
-                    resp
-                })
-                .and_then(|res| sink.success(res).map_err(Error::from))
-                .map(|_| timer.observe_duration())
-                .map_err(|e| error!("kv_prewrite failed: {:?}", e))
-                .boxed()
-        });
+        let mutations = req.take_mutations()
+            .into_iter()
+            .map(|mut x| {
+                match x.get_op() {
+                    Op::Put => Mutation::Put((Key::from_raw(x.get_key()), x.take_value())),
+                    Op::Del => Mutation::Delete(Key::from_raw(x.get_key())),
+                    Op::Lock => Mutation::Lock(Key::from_raw(x.get_key())),
+                }
+            })
+            .collect();
+        let mut options = Options::default();
+        options.lock_ttl = req.get_lock_ttl();
+        options.skip_constraint_check = req.get_skip_constraint_check();
+
+        let (cb, future) = make_callback();
+        let res = self.storage.async_prewrite(req.take_context(),
+                                              mutations,
+                                              req.take_primary_lock(),
+                                              req.get_start_version(),
+                                              options,
+                                              cb);
+        if let Err(e) = res {
+            self.send_fail_status(sink, Error::from(e), RpcStatusCode::ResourceExhausted);
+            return;
+        }
+
+        let future = future.map_err(Error::from)
+            .map(|v| {
+                let mut resp = PrewriteResponse::new();
+                if let Some(err) = extract_region_error(&v) {
+                    resp.set_region_error(err);
+                } else {
+                    resp.set_errors(RepeatedField::from_vec(extract_key_errors(v)));
+                }
+                resp
+            })
+            .and_then(|res| sink.success(res).map_err(Error::from))
+            .map(|_| timer.observe_duration())
```

Comment:
I guess no need. When _timer_ is dropped, it will call `observe_duration` automatically, so we just need to make sure the _timer_ is captured by the closure.